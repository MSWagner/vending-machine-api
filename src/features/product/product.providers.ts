import { Connection, Repository } from "typeorm";
import { Product } from "../../entities/Product.entity";

import CONFIG from "../../config";

export const productProviders = [
    {
        provide: CONFIG.database.defaultProductRepoName,
        useFactory: (connection: Connection): Repository<Product> => connection.getRepository(Product),
        inject: [CONFIG.database.defaultConnectionName]
    }
];
