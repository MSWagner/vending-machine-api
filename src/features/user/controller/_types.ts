import { ApiProperty } from "@nestjs/swagger";

export class UserPostResponse {
    @ApiProperty()
    readonly tokenType: string;

    @ApiProperty()
    readonly expiresIn: number;

    @ApiProperty({ format: "uuid" })
    refreshToken: string;

    @ApiProperty({ format: "uuid" })
    accessToken: string;
}

export class UserGetResponse {
    @ApiProperty()
    readonly username: string;

    @ApiProperty()
    readonly role: string;

    @ApiProperty()
    readonly deposit?: number;
}

export class SuccessResponse {
    @ApiProperty()
    readonly success: boolean;
}
