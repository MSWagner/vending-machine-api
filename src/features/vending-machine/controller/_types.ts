import { IsEnum, IsInt, IsUUID, Min } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";
import { IPublicProductData } from "src/entities/Product.entity";

export enum ValidCoins {
    Five = 5,
    Ten = 10,
    Twenty = 20,
    Fifty = 50,
    Hundred = 100
}

export const VALID_COINS: number[] = [
    ValidCoins.Hundred,
    ValidCoins.Fifty,
    ValidCoins.Twenty,
    ValidCoins.Ten,
    ValidCoins.Five
];

export class DepositBody {
    @IsInt()
    @IsEnum(ValidCoins)
    @ApiProperty({ type: "int", enum: ValidCoins })
    readonly coin: number;
}

export class BuyBody {
    @IsUUID()
    @ApiProperty({ type: "uuid" })
    readonly productUid: string;

    @IsInt()
    @Min(1)
    @ApiProperty({ type: "int" })
    readonly amount: number;
}

export class BuyResponse {
    @ApiProperty({ type: "product" })
    readonly product: IPublicProductData;

    @ApiProperty({ type: "int" })
    readonly spent: number;

    @ApiProperty({ type: "int", isArray: true })
    readonly change: number[];
}

export class ResetResponse {
    @ApiProperty({ type: "int", isArray: true })
    readonly change: number[];
}

export class SuccessResponse {
    @ApiProperty({ type: "boolean" })
    readonly success: boolean;
}
