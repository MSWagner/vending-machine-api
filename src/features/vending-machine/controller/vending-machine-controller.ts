import { Controller, Request, Post, Body, UseGuards, HttpException, HttpStatus, HttpCode } from "@nestjs/common";
import { AuthGuard } from "@nestjs/passport";
import { ApiCreatedResponse, ApiTags } from "@nestjs/swagger";
import { getConnection } from "typeorm";

import { Roles } from "../../auth/guards/role.decorator";
import { Product, Role } from "../../../entities/Product.entity";
import { RoleGuard } from "../../auth/guards/role.guard";

import { User } from "../../../entities/User.entity";

import { BuyBody, BuyResponse, DepositBody, ResetResponse, SuccessResponse, VALID_COINS } from "./_types";

@ApiTags("vending-machine")
@Controller({ version: "1" })
export class VendigMachineController {
    @Post("deposit")
    @HttpCode(HttpStatus.OK)
    @UseGuards(AuthGuard("bearer"), RoleGuard)
    @Roles(Role.Buyer)
    @ApiCreatedResponse({ description: "The buyer has successful deposit coins.", type: SuccessResponse })
    async deposit(@Request() req, @Body() depositBody: DepositBody): Promise<SuccessResponse> {
        const queryRunner = await getConnection().createQueryRunner();
        await queryRunner.startTransaction();

        const user = await queryRunner.manager.getRepository(User).findOne(req.user.uid);
        user.deposit += depositBody.coin;

        try {
            await await queryRunner.manager.getRepository(User).save(user);
            await queryRunner.commitTransaction();
        } catch (err) {
            await queryRunner.rollbackTransaction();
            throw new HttpException(err.message, HttpStatus.INTERNAL_SERVER_ERROR);
        } finally {
            await queryRunner.release();
        }

        return { success: true };
    }

    @Post("buy")
    @HttpCode(HttpStatus.OK)
    @UseGuards(AuthGuard("bearer"), RoleGuard)
    @Roles(Role.Buyer)
    @ApiCreatedResponse({ description: "The buyer has successful bought a product.", type: BuyResponse })
    async buy(@Request() req, @Body() buyBody: BuyBody): Promise<BuyResponse> {
        const queryRunner = await getConnection().createQueryRunner();
        await queryRunner.startTransaction();

        let product = await queryRunner.manager.getRepository(Product).findOne(buyBody.productUid);

        if (!product) {
            throw new HttpException("Product not found", HttpStatus.NOT_FOUND);
        }

        if (product.amountAvailable < 1) {
            throw new HttpException("Product is out of stock", HttpStatus.BAD_REQUEST);
        }

        if (product.amountAvailable < buyBody.amount) {
            throw new HttpException("Not enough products available", HttpStatus.BAD_REQUEST);
        }

        let user = await queryRunner.manager.getRepository(User).findOne(req.user.uid);
        const productsPrice = product.cost * buyBody.amount;

        if (user.deposit < productsPrice) {
            throw new HttpException("Insufficient deposit", HttpStatus.BAD_REQUEST);
        }

        let change = 0;

        try {
            change = user.deposit - productsPrice;
            product.amountAvailable -= buyBody.amount;
            user.deposit = 0;

            product = await queryRunner.manager.getRepository(Product).save(product);
            user = await queryRunner.manager.getRepository(User).save(user);

            await queryRunner.commitTransaction();
        } catch (err) {
            await queryRunner.rollbackTransaction();
            throw new HttpException(err.message, HttpStatus.INTERNAL_SERVER_ERROR);
        } finally {
            await queryRunner.release();
        }

        let changeCoins: number[] = [];
        if (change > 0) {
            changeCoins = this.getChangeCoins(change, VALID_COINS);
        }

        return {
            product: product.publicData,
            spent: productsPrice,
            change: changeCoins
        };
    }

    @Post("reset")
    @HttpCode(HttpStatus.OK)
    @UseGuards(AuthGuard("bearer"), RoleGuard)
    @Roles(Role.Buyer)
    @ApiCreatedResponse({ description: "The buyer has successful reset there deposit.", type: ResetResponse })
    async reset(@Request() req): Promise<ResetResponse> {
        const queryRunner = await getConnection().createQueryRunner();
        await queryRunner.startTransaction();

        const user = await queryRunner.manager.getRepository(User).findOne(req.user.uid);
        const changeCoins = this.getChangeCoins(user.deposit, VALID_COINS);
        user.deposit = 0;

        try {
            await await queryRunner.manager.getRepository(User).save(user);
            await queryRunner.commitTransaction();
        } catch (err) {
            await queryRunner.rollbackTransaction();
            throw new HttpException(err.message, HttpStatus.INTERNAL_SERVER_ERROR);
        } finally {
            await queryRunner.release();
        }

        return { change: changeCoins };
    }

    getChangeCoins(change: number, validCoins: number[]): number[] {
        const changeCoins: number[] = [];

        validCoins
            .sort((a, b) => b - a)
            .forEach((coin) => {
                while (change >= coin) {
                    changeCoins.push(coin);
                    change -= coin;
                }
            });

        return changeCoins;
    }
}
