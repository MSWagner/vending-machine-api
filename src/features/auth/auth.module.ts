import { Module } from "@nestjs/common";

import { AuthService } from "./services/auth.service";

import { LocalStrategy } from "./strategies/local.strategy";
import { BearerStrategy } from "./strategies/bearer.strategy";

import { UserModule } from "../user/user.module";

import { PassportModule } from "@nestjs/passport";

import { accessTokenProviders } from "./providers/accessTokenProviders";
import { refreshTokenProviders } from "./providers/refreshTokenProviders";
import { LoginController } from "./controller/login/login.controller";
import { RefreshTokenController } from "./controller/refresh-token/refresh-token.controller";

@Module({
    imports: [UserModule, PassportModule],
    providers: [AuthService, LocalStrategy, BearerStrategy, ...accessTokenProviders, ...refreshTokenProviders],
    controllers: [LoginController, RefreshTokenController],
    exports: [AuthService]
})
export class AuthModule {}
