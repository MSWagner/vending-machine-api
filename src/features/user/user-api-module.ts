import { Module } from "@nestjs/common";
import { UserController } from "./controller/user.controller";
import { AuthModule } from "../auth/auth.module";
import { UserModule } from "./user.module";

@Module({
    imports: [AuthModule, UserModule],
    controllers: [UserController]
})
export class UserApiModule {}
