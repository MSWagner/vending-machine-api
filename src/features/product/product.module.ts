import { Module } from "@nestjs/common";
import { ProductService } from "./services/product.service";
import { productProviders } from "./product.providers";

import { DatabaseModule } from "../../shared/database/database.module";
import { UserModule } from "../user/user.module";
import { AuthModule } from "../auth/auth.module";
import { ProductController } from "./controller/product.controller";

@Module({
    imports: [DatabaseModule, UserModule, AuthModule],
    providers: [...productProviders, ProductService],
    controllers: [ProductController],
    exports: [DatabaseModule, ProductService, ...productProviders]
})
export class ProductModule {}
