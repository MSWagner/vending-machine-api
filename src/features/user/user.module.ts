import { Module } from "@nestjs/common";
import { UserService } from "./services/user.service";
import { userProviders } from "./user.providers";

import { DatabaseModule } from "../../shared/database/database.module";

@Module({
    imports: [DatabaseModule],
    providers: [...userProviders, UserService],
    exports: [DatabaseModule, UserService, ...userProviders]
})
export class UserModule {}
