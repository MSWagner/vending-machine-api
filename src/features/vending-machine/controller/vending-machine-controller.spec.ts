import request from "supertest";
import { Test, TestingModule } from "@nestjs/testing";
import { INestApplication } from "@nestjs/common";

import { setupApp } from "../../../appSetup";

import { TestModule } from "../../../shared/test/test.module";
import { TestService } from "../../../shared/test/test.service";
import { AuthModule } from "../../../features/auth/auth.module";
import { UserModule } from "../../../features/user/user.module";
import { ProductModule } from "../../../features/product/product.module";
import { User } from "../../../entities/User.entity";
import { VendigMachineController } from "./vending-machine-controller";
import { Product } from "../../../entities/Product.entity";

import * as fixtures from "../../../shared/test/fixtures";
import { BuyBody, VALID_COINS } from "./_types";

describe("VendigMachineController", () => {
    let app: INestApplication;
    let testService: TestService;
    let vendingMachineController: VendigMachineController;

    beforeAll(async () => {
        const module: TestingModule = await Test.createTestingModule({
            controllers: [VendigMachineController],
            imports: [TestModule, AuthModule, UserModule, ProductModule]
        }).compile();

        testService = module.get<TestService>(TestService);
        vendingMachineController = module.get<VendigMachineController>(VendigMachineController);

        app = module.createNestApplication();
        await setupApp(app);
        await app.init();
    });

    beforeEach(async () => {
        await testService.reloadFixtures();
    });

    afterAll((done) => {
        testService.connectionManager.connection.close();
        done();
    });

    //// POST /deposit

    it("/POST deposit - should deposit 10 coins for user1 (buyer)", async () => {
        const body = {
            coin: 10
        };

        const user1Before = await User.findOne(fixtures.user1.uid);
        expect(user1Before.deposit).toEqual(0);

        await request(app.getHttpServer())
            .post("/api/v1/deposit")
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .send(body)
            .expect(200);

        const user1After = await User.findOne(fixtures.user1.uid);
        expect(user1After.deposit).toEqual(10);
    });

    it("/POST deposit - should deposit a second coin after the first one for user1 (buyer)", async () => {
        const body = {
            coin: 10
        };

        const user1Before = await User.findOne(fixtures.user1.uid);
        expect(user1Before.deposit).toEqual(0);

        await request(app.getHttpServer())
            .post("/api/v1/deposit")
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .send(body)
            .expect(200);

        await request(app.getHttpServer())
            .post("/api/v1/deposit")
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .send(body)
            .expect(200);

        const user1After = await User.findOne(fixtures.user1.uid);
        expect(user1After.deposit).toEqual(10 + 10);
    });

    it("/POST deposit - should throw an error for an invalid coin", async () => {
        const body = {
            coin: 150
        };

        const user1Before = await User.findOne(fixtures.user1.uid);
        expect(user1Before.deposit).toEqual(0);

        const response = await request(app.getHttpServer())
            .post("/api/v1/deposit")
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .send(body)
            .expect(400);

        const user1After = await User.findOne(fixtures.user1.uid);
        expect(user1After.deposit).toEqual(0);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("NoValidCoinError");
    });

    it("/POST deposit - should throw an error for missing a coin", async () => {
        const body = {};

        const user1Before = await User.findOne(fixtures.user1.uid);
        expect(user1Before.deposit).toEqual(0);

        const response = await request(app.getHttpServer())
            .post("/api/v1/deposit")
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .send(body)
            .expect(400);

        const user1After = await User.findOne(fixtures.user1.uid);
        expect(user1After.deposit).toEqual(0);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("MissingCoinError");
    });

    it("/POST deposit - should throw an error for a seller deposit", async () => {
        const body = {
            coin: 10
        };

        const response = await request(app.getHttpServer())
            .post("/api/v1/deposit")
            .auth(fixtures.accessTokenUser3Seller.token, { type: "bearer" })
            .send(body)
            .expect(403);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("SellerDepositError");
    });

    it("/POST deposit - should return an error for not authenticated", async () => {
        const body = {
            coin: 10
        };

        const response = await request(app.getHttpServer()).post("/api/v1/deposit").send(body).expect(401);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("NotAuthenticatedResponse");
    });

    it("/POST deposit - should return an error for a seller request", async () => {
        const body = {
            coin: 10
        };

        const response = await request(app.getHttpServer())
            .post("/api/v1/deposit")
            .auth(fixtures.accessTokenUser3Seller.token, { type: "bearer" })
            .send(body)
            .expect(403);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("SellerErrorResponse");
    });

    //// getChangeCoins

    it("getChangeCoins - should calculate the change coins for 245 change and all validCoins", async () => {
        const changeCoins = vendingMachineController.getChangeCoins(245, VALID_COINS);

        expect(changeCoins).toEqual([100, 100, 20, 20, 5]);
    });

    it("getChangeCoins - should calculate the change coins for 65 change and all validCoins", async () => {
        const changeCoins = vendingMachineController.getChangeCoins(65, VALID_COINS);

        expect(changeCoins).toEqual([50, 10, 5]);
    });

    it("getChangeCoins - should return an empty array for no change", async () => {
        const changeCoins = vendingMachineController.getChangeCoins(0, VALID_COINS);

        expect(changeCoins).toEqual([]);
    });

    //// POST /buy

    it("/POST buy - user1 should deposit 100 and buy product2 for 10 with a change of 90", async () => {
        const depositBody = {
            coin: 100
        };

        const user1Before = await User.findOne(fixtures.user1.uid);
        expect(user1Before.deposit).toEqual(0);

        await request(app.getHttpServer())
            .post("/api/v1/deposit")
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .send(depositBody)
            .expect(200);

        const user1After = await User.findOne(fixtures.user1.uid);
        expect(user1After.deposit).toEqual(100);

        const product1AmountBefore = await Product.findOne(fixtures.product2User3Seller.uid);
        expect(product1AmountBefore.amountAvailable).toEqual(1);

        const buyBody: BuyBody = {
            productUid: fixtures.product2User3Seller.uid,
            amount: 1
        };

        const buyResponse = await request(app.getHttpServer())
            .post("/api/v1/buy")
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .send(buyBody)
            .expect(200);

        expect(buyResponse.body.change).toEqual([50, 20, 20]);
        expect(buyResponse.body.spent).toEqual(10);

        const product1AmountAfter = await Product.findOne(fixtures.product2User3Seller.uid);
        expect(product1AmountAfter.amountAvailable).toEqual(0);

        const user1After2 = await User.findOne(fixtures.user1.uid);
        expect(user1After2.deposit).toEqual(0);

        const responseData = testService.replaceValues(buyResponse.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("BuyProduct2SuccessfulResponse");
    });

    it("/POST buy - user1 should deposit 200 and buy 2 product1 for 200 with a change of 0", async () => {
        const depositBody = {
            coin: 100
        };

        const user1Before = await User.findOne(fixtures.user1.uid);
        expect(user1Before.deposit).toEqual(0);

        await request(app.getHttpServer())
            .post("/api/v1/deposit")
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .send(depositBody)
            .expect(200);

        await request(app.getHttpServer())
            .post("/api/v1/deposit")
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .send(depositBody)
            .expect(200);

        const user1After = await User.findOne(fixtures.user1.uid);
        expect(user1After.deposit).toEqual(200);

        const product1AmountBefore = await Product.findOne(fixtures.product1User3Seller.uid);
        expect(product1AmountBefore.amountAvailable).toEqual(10);

        const buyBody: BuyBody = {
            productUid: fixtures.product1User3Seller.uid,
            amount: 2
        };

        const buyResponse = await request(app.getHttpServer())
            .post("/api/v1/buy")
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .send(buyBody)
            .expect(200);

        expect(buyResponse.body.change).toEqual([]);
        expect(buyResponse.body.spent).toEqual(200);

        const product1AmountAfter = await Product.findOne(fixtures.product1User3Seller.uid);
        expect(product1AmountAfter.amountAvailable).toEqual(8);

        const user1After2 = await User.findOne(fixtures.user1.uid);
        expect(user1After2.deposit).toEqual(0);

        const responseData = testService.replaceValues(buyResponse.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("BuyProduct1TwiceSuccessfulResponse");
    });

    it("/POST buy - should throw an error for not enough products available", async () => {
        const depositBody = {
            coin: 100
        };

        const user1Before = await User.findOne(fixtures.user1.uid);
        expect(user1Before.deposit).toEqual(0);

        await request(app.getHttpServer())
            .post("/api/v1/deposit")
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .send(depositBody)
            .expect(200);

        const user1After = await User.findOne(fixtures.user1.uid);
        expect(user1After.deposit).toEqual(100);

        const product1AmountBefore = await Product.findOne(fixtures.product2User3Seller.uid);
        expect(product1AmountBefore.amountAvailable).toEqual(1);

        const buyBody: BuyBody = {
            productUid: fixtures.product2User3Seller.uid,
            amount: 2
        };

        const buyResponse = await request(app.getHttpServer())
            .post("/api/v1/buy")
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .send(buyBody)
            .expect(400);

        const product1AmountAfter = await Product.findOne(fixtures.product2User3Seller.uid);
        expect(product1AmountAfter.amountAvailable).toEqual(1);

        const user1After2 = await User.findOne(fixtures.user1.uid);
        expect(user1After2.deposit).toEqual(100);

        const responseData = testService.replaceValues(buyResponse.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("NotEnoughProductsAvailableResponse");
    });

    it("/POST buy - should throw an error for not enough deposit", async () => {
        const depositBody = {
            coin: 100
        };

        const user1Before = await User.findOne(fixtures.user1.uid);
        expect(user1Before.deposit).toEqual(0);

        await request(app.getHttpServer())
            .post("/api/v1/deposit")
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .send(depositBody)
            .expect(200);

        const user1After = await User.findOne(fixtures.user1.uid);
        expect(user1After.deposit).toEqual(100);

        const product1AmountBefore = await Product.findOne(fixtures.product1User3Seller.uid);
        expect(product1AmountBefore.amountAvailable).toEqual(10);

        const buyBody: BuyBody = {
            productUid: fixtures.product1User3Seller.uid,
            amount: 2
        };

        const buyResponse = await request(app.getHttpServer())
            .post("/api/v1/buy")
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .send(buyBody)
            .expect(400);

        const product1AmountAfter = await Product.findOne(fixtures.product1User3Seller.uid);
        expect(product1AmountAfter.amountAvailable).toEqual(10);

        const user1After2 = await User.findOne(fixtures.user1.uid);
        expect(user1After2.deposit).toEqual(100);

        const responseData = testService.replaceValues(buyResponse.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("NotEnoughDepositResponse");
    });

    it("/POST buy - should throw an error for product not found", async () => {
        const depositBody = {
            coin: 100
        };

        const user1Before = await User.findOne(fixtures.user1.uid);
        expect(user1Before.deposit).toEqual(0);

        await request(app.getHttpServer())
            .post("/api/v1/deposit")
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .send(depositBody)
            .expect(200);

        const user1After = await User.findOne(fixtures.user1.uid);
        expect(user1After.deposit).toEqual(100);

        const product1AmountBefore = await Product.findOne(fixtures.product1User3Seller.uid);
        expect(product1AmountBefore.amountAvailable).toEqual(10);

        const buyBody: BuyBody = {
            productUid: "b94f76b3-c60c-43b6-b479-1dbc37bb805d",
            amount: 2
        };

        const buyResponse = await request(app.getHttpServer())
            .post("/api/v1/buy")
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .send(buyBody)
            .expect(404);

        const product1AmountAfter = await Product.findOne(fixtures.product1User3Seller.uid);
        expect(product1AmountAfter.amountAvailable).toEqual(10);

        const user1After2 = await User.findOne(fixtures.user1.uid);
        expect(user1After2.deposit).toEqual(100);

        const responseData = testService.replaceValues(buyResponse.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("ProductNotFoundResponse");
    });

    it("/POST buy - should return an error for not authenticated", async () => {
        const buyBody: BuyBody = {
            productUid: "b94f76b3-c60c-43b6-b479-1dbc37bb805d",
            amount: 2
        };

        const response = await request(app.getHttpServer()).post("/api/v1/buy").send(buyBody).expect(401);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("NotAuthenticatedResponse");
    });

    it("/POST buy - should return an error for a seller request", async () => {
        const buyBody: BuyBody = {
            productUid: "b94f76b3-c60c-43b6-b479-1dbc37bb805d",
            amount: 2
        };

        const response = await request(app.getHttpServer())
            .post("/api/v1/buy")
            .auth(fixtures.accessTokenUser3Seller.token, { type: "bearer" })
            .send(buyBody)
            .expect(403);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("SellerErrorResponse");
    });

    //// POST /reset

    it("/POST reset - should reset the deposit of user1", async () => {
        const body = {
            coin: 10
        };

        const user1Before = await User.findOne(fixtures.user1.uid);
        expect(user1Before.deposit).toEqual(0);

        await request(app.getHttpServer())
            .post("/api/v1/deposit")
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .send(body)
            .expect(200);

        const user1After = await User.findOne(fixtures.user1.uid);
        expect(user1After.deposit).toEqual(10);

        const response = await request(app.getHttpServer())
            .post("/api/v1/reset")
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .expect(200);

        expect(response.body.change).toEqual([10]);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("ResetDeposit10Response");
    });

    it("/POST reset - should return an empty array for no deposit", async () => {
        const user1Before = await User.findOne(fixtures.user1.uid);
        expect(user1Before.deposit).toEqual(0);

        const response = await request(app.getHttpServer())
            .post("/api/v1/reset")
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .expect(200);

        expect(response.body.change).toEqual([]);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("ResetDeposit0Response");
    });

    it("/POST reset - should return an error for not authenticated", async () => {
        const response = await request(app.getHttpServer()).post("/api/v1/reset").expect(401);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("NotAuthenticatedResponse");
    });

    it("/POST reset - should return an error for a seller request", async () => {
        const response = await request(app.getHttpServer())
            .post("/api/v1/reset")
            .auth(fixtures.accessTokenUser3Seller.token, { type: "bearer" })
            .expect(403);

        const responseData = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(responseData).toMatchSnapshot("SellerErrorResponse");
    });
});
