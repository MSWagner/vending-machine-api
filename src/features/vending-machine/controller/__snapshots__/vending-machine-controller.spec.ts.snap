// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VendigMachineController /POST buy - should return an error for a seller request: SellerErrorResponse 1`] = `
{
  "error": "Forbidden",
  "message": "Forbidden resource",
  "statusCode": 403,
}
`;

exports[`VendigMachineController /POST buy - should return an error for not authenticated: NotAuthenticatedResponse 1`] = `
{
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`VendigMachineController /POST buy - should throw an error for not enough deposit: NotEnoughDepositResponse 1`] = `
{
  "message": "Insufficient deposit",
  "statusCode": 400,
}
`;

exports[`VendigMachineController /POST buy - should throw an error for not enough products available: NotEnoughProductsAvailableResponse 1`] = `
{
  "message": "Not enough products available",
  "statusCode": 400,
}
`;

exports[`VendigMachineController /POST buy - should throw an error for product not found: ProductNotFoundResponse 1`] = `
{
  "message": "Product not found",
  "statusCode": 404,
}
`;

exports[`VendigMachineController /POST buy - user1 should deposit 100 and buy product2 for 10 with a change of 90: BuyProduct2SuccessfulResponse 1`] = `
{
  "change": [
    50,
    20,
    20,
  ],
  "product": {
    "amountAvailable": 0,
    "cost": 10,
    "createdAt": "createdAt",
    "productName": "product2",
    "uid": "uid",
    "updatedAt": "updatedAt",
  },
  "spent": 10,
}
`;

exports[`VendigMachineController /POST buy - user1 should deposit 200 and buy 2 product1 for 200 with a change of 0: BuyProduct1TwiceSuccessfulResponse 1`] = `
{
  "change": [],
  "product": {
    "amountAvailable": 8,
    "cost": 100,
    "createdAt": "createdAt",
    "productName": "product1",
    "uid": "uid",
    "updatedAt": "updatedAt",
  },
  "spent": 200,
}
`;

exports[`VendigMachineController /POST deposit - should return an error for a seller request: SellerErrorResponse 1`] = `
{
  "error": "Forbidden",
  "message": "Forbidden resource",
  "statusCode": 403,
}
`;

exports[`VendigMachineController /POST deposit - should return an error for not authenticated: NotAuthenticatedResponse 1`] = `
{
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`VendigMachineController /POST deposit - should throw an error for a seller deposit: SellerDepositError 1`] = `
{
  "error": "Forbidden",
  "message": "Forbidden resource",
  "statusCode": 403,
}
`;

exports[`VendigMachineController /POST deposit - should throw an error for an invalid coin: NoValidCoinError 1`] = `
{
  "error": "Bad Request",
  "message": [
    "coin must be a valid enum value",
  ],
  "statusCode": 400,
}
`;

exports[`VendigMachineController /POST deposit - should throw an error for missing a coin: MissingCoinError 1`] = `
{
  "error": "Bad Request",
  "message": [
    "coin must be a valid enum value",
    "coin must be an integer number",
  ],
  "statusCode": 400,
}
`;

exports[`VendigMachineController /POST reset - should reset the deposit of user1: ResetDeposit10Response 1`] = `
{
  "change": [
    10,
  ],
}
`;

exports[`VendigMachineController /POST reset - should return an empty array for no deposit: ResetDeposit0Response 1`] = `
{
  "change": [],
}
`;

exports[`VendigMachineController /POST reset - should return an error for a seller request: SellerErrorResponse 1`] = `
{
  "error": "Forbidden",
  "message": "Forbidden resource",
  "statusCode": 403,
}
`;

exports[`VendigMachineController /POST reset - should return an error for not authenticated: NotAuthenticatedResponse 1`] = `
{
  "message": "Unauthorized",
  "statusCode": 401,
}
`;
